src/ngraph/pass/liveness.cpp

```diff
@@ -14,117 +14,173 @@
+        if (auto constant_node = dynamic_cast<const op::Constant*>(n))
         {
             for (size_t i = 0; i < constant_node->get_output_size(); ++i)
             {
-                descriptor::Tensor& tensor = constant_node->get_output_tensor(i);
-                persistent_tensors.insert(&tensor);
+                descriptor::Tensor& t = constant_node->get_output_tensor(i);
+                tensors.insert(&t);
             }
         }
     }
 
-    unordered_set<descriptor::Tensor*> currently_live;
-    for (auto it = ops.rbegin(); it != ops.rend(); it++)
+    // Tensors to provide the parameters of the graph function...
+    for (const shared_ptr<op::Parameter>& n : f.get_parameters())
     {
-        shared_ptr<Node> node = *it;
-        node->liveness_new_list.clear();
-        node->liveness_free_list.clear();
-        unordered_set<descriptor::Tensor*> input_tensor_decls;
-        for (descriptor::Input& input_decl : node->get_inputs())
+        for (size_t i = 0; i < n->get_output_size(); ++i)
         {
-            descriptor::Tensor& tensor = input_decl.get_tensor();
-            if (persistent_tensors.find(&tensor) == persistent_tensors.end())
-            {
-                input_tensor_decls.insert(&tensor);
-            }
+            descriptor::Tensor& t = n->get_output_tensor(i);
+            tensors.insert(&t);
         }
+    }
 
-        unordered_set<descriptor::Tensor*> output_tensor_decls;
-        for (size_t i = 0; i < node->get_output_size(); ++i)
+    // Tensors to receive the results of the graph function...
+    for (const shared_ptr<op::Result>& n : f.get_results())
+    {
+        for (size_t i = 0; i < n->get_output_size(); ++i)
         {
-            descriptor::Tensor& tensor = node->get_output_tensor(i);
-            if (persistent_tensors.find(&tensor) == persistent_tensors.end())
-            {
-                output_tensor_decls.insert(&tensor);
-            }
+            descriptor::Tensor& t = n->get_output_tensor(i);
+            tensors.insert(&t);
         }
+    }
+
+    return tensors;
+}
+
+bool pass::Liveness::run_on_function(shared_ptr<ngraph::Function> function)
+{
+    //------------------------------------------------------------------------------------------------
+    // ALGORITHM OVERVIEW:
+    // This pass uses a backwards-dataflow analysis, very similar to what one would find in textbook
+    // descriptions of liveness analysis.
+    // Note that we visit the ops in *reverse* execution order.  So we visit the final use (if there
+    // is one) of each non-persistent tensor before visiting the op that generates that tensor.
+    //------------------------------------------------------------------------------------------------
+
+    ngraph::Function& f = *(function.get());
+
+    // FIXME: Ideally this should be computed just once and passed in to the function...
+    const vector<Node*> ops_execution_order = get_ordered_ops_vector(f);
```

Comment:
nearly every single optimization we have invalidates this list, that's why it's always recomputed. if `replace_node` or `replace_output` had access to `Function` we could invalidate it automatically and use a cached version otherwise

Reply:
> nearly every single optimization we have invalidates this list, that's why it's always recomputed.

So following that logic, and optimization which alters that list _also_ invalidates the liveness analysis, right?

If so, I wonder if we should make that more obvious to pass- and/or back-end- developers?